### PR TITLE
gh-123213: Fixed xml.etree.ElementTree.Element.extend and assignment to no longer hide exceptions

### PR DESCRIPTION
**Pull request by bharel**
Backport to 3.12 please.

Not formatting the exception message with the type (i.e "expected iterable not '...'") per `PySequence_Fast` standard usage across all modules.

Apparently this bug exists for 15 years ☹️ 

Typeshed already expects an Iterable for `.extend()` and `__setitem__`, so it does not require an update ❤️ 

* Issue: gh-123213

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-08-22-09-37-48.gh-issue-123213.owmXnP.rst
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-08-22-09-37-48.gh-issue-123213.owmXnP.rst]:
```suggestion
:meth:`xml.etree.ElementTree.Element.extend` and ``Element`` assignment no
longer hide the internal exception if an erronous generator is passed. Patch
by Bar Harel.
```

File: Modules/_elementtree.c
Comment (picnixz) [inline Modules/_elementtree.c:1216]: Ideally we want to keep the message about the incorrect type.

Comment (bharel) [reply Modules/_elementtree.c:1216]: Per what I wrote in the PR, I've standardized it according to all other usages of pysequence_fast.
If you wish to retain the message it will require formatting the error string on every invocation regardless of the input or require changing all instances of pysequence_fast which is out of scope.

File: Misc/NEWS.d/next/Library/2024-08-22-09-37-48.gh-issue-123213.owmXnP.rst
Comment (bharel) [reply Misc/NEWS.d/next/Library/2024-08-22-09-37-48.gh-issue-123213.owmXnP.rst]: Thanks :-)

File: Modules/_elementtree.c
Comment (picnixz) [reply Modules/_elementtree.c:1216]: There is something I'm actually wondering. Should we actually accept any iterable or just sequences? because here the message says "expected sequence" and not "expected iterable". 

Anyway, your argument is sound and consistent with other occurrences of `PySequence_Fast` (except for the one in dictobject but this one is special I think)

Comment (bharel) [reply Modules/_elementtree.c:1216]: I believe any iterable, per all other usages and implementation of PySequence, the typeshed and the general commonplace usage of `.extend()` methods.

I haven't seen anywhere in the code where a sequence is required before the implicit conversion to PySequence. The pure Python implementation also accepts arbitrary iterables, which currently means there's a deviation between the Python and C code in case of errors.

IIRC the docs do state sequence but I can change the docs to be more lenient. It is very common for it to write sequences in places where any iterables are accepted so I haven't touched it.

As it is more permissive, no bugs are relevant except for potential inheritance specifically from the C implementation breaking LSP, but the interface according to the shed is any Iterable anyway.

All in all, yes, I believe the bug is a bug :-)

Thanks for reviewing btw 🐍

Comment (bharel) [reply Modules/_elementtree.c:1216]: I double checked now, and as the biggest relevant external package that is also a drop-in replacement, lxml also accepts an iterable. It further hints towards an iterable being the expected behavior.

Let's say like this - if we change it to a sequence only, we're going to break a lot of code, and we'll require a more specific type for no apparent reason.

If we change it to an arbitrary iterable, the odds of breaking any code are astronomically low, and the behavior will be as expected by code in the wild.

I also had that doubt at first that maybe I wrote bad code, but it seems like there's a bug here.

Comment (picnixz) [reply Modules/_elementtree.c:1216]: Thanks for the investigation. I don't mind amending the docs since we are anyway accepting any iterable (and not just sequences). You'll need to wait for a core dev to review your PR though since I don't have commit permissions.

Comment (bharel) [reply Modules/_elementtree.c:1216]: Amended the docs per your suggestion.

Haven't added a versionchanged tag as it's technically a fix and clarification. We accepted iterables all along, just with a bug in case of an exception. 

Thank you for your review and suggestions.
If you believe it is ready you can resolve the comment and accept the review so the core-reviewer tag will be added by the bot 🥂

File: Lib/test/test_xml_etree.py
Comment (picnixz) [inline Lib/test/test_xml_etree.py:2432]: Just one more test case maybe: use a generator of the form `(f() for i in range(2))` where `f` raises `TypeError("hello")`. Then check that this TypeError message is the correct one.

File: Lib/test/test_xml_etree.py
Comment (picnixz) [inline Lib/test/test_xml_etree.py:3774]: Ditto.

File: Misc/NEWS.d/next/Library/2024-08-22-09-37-48.gh-issue-123213.owmXnP.rst
Comment (picnixz) [inline Misc/NEWS.d/next/Library/2024-08-22-09-37-48.gh-issue-123213.owmXnP.rst]: "Element assignment" should either link to the corresponding method in the docs, or `Element` should be linked.

File: Modules/_elementtree.c
Comment (picnixz) [inline Modules/_elementtree.c]:
```suggestion
    seq = PySequence_Fast(elements, "'elements' must be an iterable");
```

When you extend a list, it says `TypeError: "int" object is not iterable`. Since we are removing the type information, I'd suggest at least having the parameter name in the message.

**Review by picnixz — COMMENTED** [review]
As I said previously, your argument is sound. However, one could argue that `list.extend` raises a TypeError with the incorrect type in the error message. I don't mind either suggestions (technically, the message *could* include the type information but you'll probably need to play with some `PyErr_ExceptionMatches` or remove the usage of `PySequence_Fast` and do something different). I personally would like feedback from someone else so that we could arrive at the best solution both for the user and for the maintainability.

File: Modules/_elementtree.c
Comment (picnixz) [reply Modules/_elementtree.c:1216]: I'll let this one opened so that other devs could discuss here.

File: Lib/test/test_xml_etree.py
Comment (bharel) [reply Lib/test/test_xml_etree.py:2432]: Good idea, that's also why I didn't want to play with PyErr_ExceptionMatches.

**Comment by bharel** [comment]
@eliben our xml expert 🙇 
Quick one for the morning 😄

**Comment by picnixz** [comment]
I'm not sure that eli is still active/available but we can ask @serhiy-storchaka for this one.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

**Pull request merged**
